File: scripts/gesture-probe.py
#!/usr/bin/env python3
"""Drive `src/bin/gesture_probe.rs` over Rinch's debug IPC and print what fired.

Card C6's spike. The probe app writes every pointer event into one text node;
this connects to the debug server, synthesises presses and moves through the
*real* input path (`mouse_down` / `mouse_move` / `mouse_up` all go through
`RinchApp::handle_event`, exactly as a physical mouse does), then reads that
node back.

No coordinate is ever read off a picture: every one comes from a
`query_selector` layout box, which is what `docs/NOTES.md`'s HiDPI warning asks
for.

    cargo run --release --features devtools --bin gesture_probe &
    scripts/gesture-probe.py
"""

import json
import os
import socket
import struct
import sys
import time

APP = "gesture_probe"


def discover():
    """The port of the newest live gesture_probe, from ~/.rinch/debug."""
    d = os.path.join(os.path.expanduser("~"), ".rinch", "debug")
    best = None
    for name in os.listdir(d):
        try:
            entry = json.load(open(os.path.join(d, name)))
        except Exception:
            continue
        if entry.get("app_name") != APP:
            continue
        try:
            os.kill(entry["pid"], 0)
        except OSError:
            continue  # stale file from a dead run
        if best is None or entry["pid"] > best["pid"]:
            best = entry
    if best is None:
        sys.exit(f"no live {APP} found in {d} — is it running with --features devtools?")
    return best["port"]


class Client:
    def __init__(self, port):
        self.sock = socket.create_connection(("127.0.0.1", port), timeout=10)
        self.id = 0
        self._frame(json.dumps({"protocol": "rinch-debug", "version": 1}).encode())
        hs = json.loads(self._read())
        assert hs.get("protocol") == "rinch-debug", hs

    def _frame(self, data):
        self.sock.sendall(struct.pack(">I", len(data)) + data)

    def _read(self):
        (n,) = struct.unpack(">I", self._recv_exact(4))
        return self._recv_exact(n)

    def _recv_exact(self, n):
        buf = b""
        while len(buf) < n:
            chunk = self.sock.recv(n - len(buf))
            if not chunk:
                raise EOFError("debug server closed the connection")
            buf += chunk
        return buf

    def cmd(self, method, **params):
        self.id += 1
        # `wait_frame` and friends are unit variants: serde rejects an empty
        # params map where it expects no field at all.
        req = {"id": self.id, "method": method}
        if params:
            req["params"] = params
        self._frame(json.dumps(req).encode())
        resp = json.loads(self._read())
        if resp.get("type") == "error":
            raise RuntimeError(f"{method}: {resp['message']}")
        return resp.get("data")

    # ── conveniences ────────────────────────────────────────────────────

    def select(self, selector):
        return self.cmd("query_selector", selector=selector)

    def one(self, selector):
        got = self.select(selector)
        if not got:
            sys.exit(f"nothing matched {selector}")
        return got[0]

    def box(self, node):
        a = node["absolute"]
        return a["x"], a["y"], a["width"], a["height"]

    def centre(self, node):
        x, y, w, h = self.box(node)
        return x + w / 2, y + h / 2

    def log(self):
        return self.one(".probe-log")["text_content"]

    def reset(self):
        x, y = self.centre(self.one(".probe-reset"))
        self.cmd("click", x=x, y=y)
        self.cmd("wait_frame")

    def drag(self, x0, y0, x1, y1, steps=8, settle=0.02):
        """A press, `steps` moves, a release — the way a hand does it."""
        self.cmd("mouse_move", x=x0, y=y0)
        self.cmd("mouse_down", x=x0, y=y0)
        for i in range(1, steps + 1):
            t = i / steps
            self.cmd("mouse_move", x=x0 + (x1 - x0) * t, y=y0 + (y1 - y0) * t)
            time.sleep(settle)
        self.cmd("mouse_up", x=x1, y=y1)
        self.cmd("wait_frame")


def banner(title):
    print(f"\n\033[1m── {title} " + "─" * max(0, 60 - len(title)) + "\033[0m")


def main():
    c = Client(discover())
    rows = c.select(".probe-row")
    handles = c.select(".probe-handle")
    lst = c.one(".probe-list")
    print(f"list box        {c.box(lst)}")
    print(f"{len(rows)} rows, first at {c.box(rows[0])}, second at {c.box(rows[1])}")

    # ── 1. Drag the handle of row 1 down over row 3 ─────────────────────
    banner("1. drag the handle downward, three rows")
    c.reset()
    hx, hy = c.centre(handles[1])
    _, ty = c.centre(rows[4])
    c.drag(hx, hy, hx, ty)
    print(c.log())

    # ── 2. Horizontal swipe on a row body (no handle) ───────────────────
    banner("2. swipe left across a row body, 120px")
    c.reset()
    rx, ry, rw, rh = c.box(rows[6])
    start_x = rx + rw - 20
    c.drag(start_x, ry + rh / 2, start_x - 120, ry + rh / 2)
    print(c.log())

    # ── 3. Vertical drag on a row body — the gesture a finger would use
    #       to scroll the list. On the desktop this is not scrolling at
    #       all; the point is what the row's own handlers see.
    banner("3. vertical drag across a row body, 200px")
    c.reset()
    rx, ry, rw, rh = c.box(rows[6])
    c.drag(rx + rw / 2, ry + rh / 2, rx + rw / 2, ry + rh / 2 - 200)
    print(c.log())

    # ── 4. Does the wheel — the only thing that scrolls this box on the
    #       desktop, and what Android turns a finger-drag into — reach the
    #       row, and does the list actually move? `delta_y` is added to the
    #       offset, so scrolling *down* is positive.
    banner("4. wheel over a row")
    c.reset()
    rx, ry, rw, rh = c.box(rows[6])
    c.cmd("scroll", x=rx + rw / 2, y=ry + rh / 2, delta_x=0.0, delta_y=180.0)
    c.cmd("wait_frame")
    print(c.log())
    print("row 0 box after scroll:", c.box(c.select(".probe-row")[0]))

    # ── 4b. …and does a drag still land on the right row once the list has
    #        been scrolled? Layout boxes are re-read after the scroll.
    banner("4b. drag a handle after scrolling")
    c.reset()
    rows2 = c.select(".probe-row")
    handles2 = c.select(".probe-handle")
    hx, hy = c.centre(handles2[6])
    _, ty = c.centre(rows2[9])
    print(f"handle 6 at {hx:.0f},{hy:.0f}; row 9 centre y {ty:.0f}")
    c.drag(hx, hy, hx, ty)
    print(c.log())
    c.cmd("scroll", x=rx + rw / 2, y=ry + rh / 2, delta_x=0.0, delta_y=-1000.0)
    c.cmd("wait_frame")

    # ── 5. A plain tap must still be a tap once handles are draggable ───
    banner("5. tap a row body")
    c.reset()
    x, y = c.centre(rows[2])
    c.cmd("click", x=x, y=y)
    c.cmd("wait_frame")
    print(c.log())

    # ── 6. A plain tap on the handle itself ─────────────────────────────
    banner("6. tap the drag handle")
    c.reset()
    x, y = c.centre(handles[2])
    c.cmd("click", x=x, y=y)
    c.cmd("wait_frame")
    print(c.log())


if __name__ == "__main__":
    main()

File: src/bin/gesture_probe.rs
//! Spike harness for card C6 — do drag and swipe actually work?
//!
//! Card C6 asks for reorder-by-drag-handle and swipe-left-to-remove inside a
//! setlist. Three separate pointer/paint faults (K10, K11, K14) have already
//! been found in this framework by this project, so the gestures are proven
//! here, in the smallest thing that can hold them, before any setlist UI is
//! built on top of them.
//!
//! It reproduces the exact shape the real screen has and nothing else: rows
//! inside an `overflow-y: auto` scroll box, each with a drag handle. The two
//! questions are
//!
//!   (a) does a drag fire its events in the order the feature needs, with
//!       coordinates that can be turned into "row 3 goes above row 1"; and
//!   (b) can a horizontal swipe on a row be told apart from a vertical scroll
//!       of the list the row is sitting in.
//!
//! Every handler appends to a single text node (`.probe-log`), so the whole
//! event trace can be read back over the `--features devtools` IPC with
//! `query_selector` + `get_text_content` — no screenshots, and therefore none
//! of the HiDPI coordinate trouble `docs/NOTES.md` warns about. Drive it with
//! `scripts/gesture-probe.py`.
//!
//! ```text
//! cargo run --release --features devtools --bin gesture_probe
//! ```
//!
//! ## What it found
//!
//! On the **desktop** backend, both gestures work and the trace is unambiguous:
//!
//! ```text
//! mousedown r1 @29,415 elem 12,387 467x56 | dragstart r1 @29,436
//!   | dragenter r2 | dragover r2 @29,457 | … | dragenter r4
//!   | dragover r4 @29,583 | mouseup r4 | drop r4 @29,583 | dragend r1
//! ```
//!
//! — the source row from `dragstart`, the target row from `drop`, which is
//! exactly "row 1 goes after row 4". It still works with the list scrolled
//! (layout boxes carry the scroll offset), a plain tap on a `draggable` handle
//! is still a tap, and a horizontal drag across a row body reports its own
//! delta (`mouseup r6 @339,695 d=-120,0`). Nothing competes with it, because a
//! mouse drag does not scroll an `overflow-y: auto` box on the desktop at all —
//! only the wheel does, and the wheel never reaches the row's handlers.
//!
//! On **Android** neither gesture can fire, and the reason is in the shell
//! rather than in anything an app can write. See `docs/ANDROID.md`, "Touch on
//! Android is a tap and a scroll, and nothing else". The short version: `MouseDown` is
//! emitted only at finger-*up*, immediately followed by `MouseUp` at the same
//! point, and only for a finger that never travelled 8px — so the runtime's
//! pending drag is created and consumed in one batch and becomes a click. A
//! finger that moves produces `MouseWheel` deltas addressed at the scroll
//! container and no app-visible event whatsoever, including none when it lifts.
//!
//! Which is why `src/screens/setlist_detail.rs` reorders with buttons. Keep
//! this probe until the Android recogniser is fixed upstream: it is the thing
//! that turns "gestures feel broken on the phone" back into a five-minute
//! answer.

// Same rsx! lint artifact as main.rs: bindings used inside generated closures
// are reported as unused.
#![allow(unused_variables)]

use rinch::prelude::*;
use std::cell::RefCell;
use std::rc::Rc;

/// The trace every handler writes into. A `Vec<String>` behind a signal so the
/// log node re-renders, plus a plain counter of the mousemoves that were
/// folded away — a swipe produces dozens and the interesting part is the first
/// and the last, not the middle.
#[derive(Clone, Copy)]
struct Trace {
    lines: Signal<Vec<String>>,
    moves: Signal<u32>,
}

impl Trace {
    fn push(self, line: String) {
        self.lines.update(|l| l.push(line));
    }

    /// A mousemove, recorded in full for the first few of a gesture and only
    /// counted after that.
    fn mouse_move(self, row: usize, x: f32, y: f32) {
        let n = self.moves.get();
        self.moves.set(n + 1);
        if n < 3 {
            self.push(format!("mousemove r{row} @{x:.0},{y:.0}"));
        }
    }

    fn reset(self) {
        self.lines.set(Vec::new());
        self.moves.set(0);
    }
}

const ROWS: usize = 20;

/// A pointer position, remembered across handler calls so a swipe can measure
/// itself. `Rc<RefCell<..>>` rather than a signal: nothing renders from it, and
/// writing a signal from inside a mousemove would re-render the list on every
/// pixel of a drag.
type Press = Rc<RefCell<Option<(f32, f32, usize)>>>;

#[component]
fn app() -> NodeHandle {
    let trace = Trace {
        lines: Signal::new(vec!["ready".to_string()]),
        moves: Signal::new(0),
    };
    let press: Press = Rc::new(RefCell::new(None));

    rsx! {
        div {
            style: "display: flex; flex-direction: column; height: 100vh; \
                    background: #FBF7F0; color: #1C1917; font-size: 12px;",

            // The trace, read back over the IPC. One text node on purpose:
            // `get_text_content` on a single node is the whole log.
            div {
                style: "flex-shrink: 0; height: 300px; overflow: hidden; padding: 8px; \
                        background: #F1E9DC; border-bottom: 2px solid #B54724; \
                        font-family: monospace; line-height: 1.35;",
                div {
                    class: "probe-log",
                    {move || {
                        let mut lines = trace.lines.get();
                        let moves = trace.moves.get();
                        if moves > 3 {
                            lines.push(format!("(+{} more mousemove)", moves - 3));
                        }
                        lines.join(" | ")
                    }}
                }
            }

            div {
                class: "probe-reset",
                onclick: move || trace.reset(),
                style: "flex-shrink: 0; padding: 6px 8px; background: #1C1917; color: #FBF7F0;",
                "reset trace"
            }

            // The shape that matters: rows inside a box that scrolls in the
            // *other* axis from the swipe. Question (b) lives right here.
            div {
                class: "probe-list",
                onscroll: move |top: f64| trace.push(format!("scroll {top:.0}")),
                style: "flex: 1; min-height: 0; overflow-y: auto; padding: 0 12px;",

                for i in 0..ROWS {
                    // Each `for` body is its own re-running closure, so it can
                    // only capture `Copy` values and owned clones (docs/NOTES.md).
                    let press = press.clone();
                    let press_move = press.clone();
                    let press_up = press.clone();
                    div {
                        key: i,
                        class: "probe-row",
                        style: "display: flex; align-items: center; gap: 10px; height: 56px; \
                                border-bottom: 1px solid #E7DFD4;",

                        onmousedown: move || {
                            let c = get_click_context();
                            *press.borrow_mut() = Some((c.mouse_x, c.mouse_y, i));
                            trace.push(format!(
                                "mousedown r{i} @{:.0},{:.0} elem {:.0},{:.0} {:.0}x{:.0}",
                                c.mouse_x, c.mouse_y,
                                c.element_x, c.element_y, c.element_width, c.element_height
                            ));
                        },
                        onmousemove: move || {
                            let c = get_click_context();
                            // Only trace moves that belong to a press that
                            // started on this row — a bare hover over the list
                            // would otherwise bury the gesture.
                            if press_move.borrow().is_some() {
                                trace.mouse_move(i, c.mouse_x, c.mouse_y);
                            }
                        },
                        onmouseup: move || {
                            let c = get_click_context();
                            let started = press_up.borrow_mut().take();
                            match started {
                                Some((sx, sy, _)) => trace.push(format!(
                                    "mouseup r{i} @{:.0},{:.0} d={:.0},{:.0}",
                                    c.mouse_x, c.mouse_y, c.mouse_x - sx, c.mouse_y - sy
                                )),
                                None => trace.push(format!("mouseup r{i} (no press)")),
                            }
                        },

                        // Drop target, so a drag over this row reports itself.
                        ondragenter: move || trace.push(format!("dragenter r{i}")),
                        ondragover: move || {
                            let c = get_click_context();
                            trace.push(format!("dragover r{i} @{:.0},{:.0}", c.mouse_x, c.mouse_y));
                        },
                        ondrop: move || {
                            let c = get_click_context();
                            trace.push(format!("drop r{i} @{:.0},{:.0}", c.mouse_x, c.mouse_y));
                        },

                        // The drag handle. `draggable="true"` is what puts the
                        // runtime into a pending drag on mousedown.
                        div {
                            class: "probe-handle",
                            draggable: "true",
                            ondragstart: move || {
                                let c = get_click_context();
                                trace.push(format!(
                                    "dragstart r{i} @{:.0},{:.0}", c.mouse_x, c.mouse_y
                                ));
                            },
                            ondragmove: move || {
                                let c = get_click_context();
                                trace.mouse_move(i, c.mouse_x, c.mouse_y);
                            },
                            ondragend: move || {
                                let c = get_click_context();
                                trace.push(format!(
                                    "dragend r{i} @{:.0},{:.0}", c.mouse_x, c.mouse_y
                                ));
                            },
                            style: "width: 34px; height: 40px; flex-shrink: 0; background: #E7DFD4; \
                                    display: flex; align-items: center; justify-content: center;",
                            "::"
                        }

                        div { style: "flex: 1;", {format!("row {i}")} }
                    }
                }
            }
        }
    }
}

#[cfg(not(target_os = "android"))]
fn main() {
    run("gesture_probe", 393, 852, app);
}

// Desktop-only, like `probe.rs`. Kept compilable for Android so a whole-crate
// cross build still resolves.
#[cfg(target_os = "android")]
fn main() {}
